Mathlib/Data/Matroid/Constructions.lean

```diff
@@ -0,0 +1,245 @@
+
+## Main definitions
+
+* `emptyOn α` is the matroid on `α` with empty ground set.
+
+For `E : Set α`, ...
+
+* `loopyOn E` is the matroid on `E` whose elements are all loops, or equivalently in which `∅`
+  is the only base.
+* `freeOn E` is the 'free matroid' whose ground set `E` is the only base.
+* For `I ⊆ E`, `uniqueBaseOn I E` is the matroid with ground set `E` in which `I` is the only base.
+
+## Implementation details
+
+To avoid the tedious process of certifying the matroid axioms for each of these easy examples,
+we bootstrap the definitions starting with `emptyOn α` (which `simp` can prove is a matroid)
+and then construct the other examples using duality and restriction.
+
+-/
+
+variable {α : Type*} {M : Matroid α} {E B I X R J : Set α}
+
+namespace Matroid
+
+open Set
+
+section EmptyOn
+
+/-- The `Matroid α` with empty ground set-/
+def emptyOn (α : Type*) : Matroid α :=
+  Matroid.ofBaseOfFinite finite_empty (· = ∅) ⟨_, rfl⟩ (by rintro _ _ rfl; simp) (by simp)
+
+@[simp] theorem emptyOn_ground : (emptyOn α).E = ∅ := rfl
+
+@[simp] theorem emptyOn_base_iff : (emptyOn α).Base B ↔ B = ∅ := by
+  simp [emptyOn]
+
+@[simp] theorem emptyOn_indep_iff : (emptyOn α).Indep B ↔ B = ∅ := by
+  simp [indep_iff_subset_base, subset_empty_iff]
+
+@[simp] theorem ground_eq_empty_iff : (M.E = ∅) ↔ M = emptyOn α := by
+  refine' ⟨fun h ↦ eq_of_base_iff_base_forall (by simp [h]) _, fun h ↦ by simp [h]⟩
+  simp only [h, subset_empty_iff, emptyOn_base_iff, forall_eq, iff_true]
+  obtain ⟨B', hB'⟩ := M.exists_base
+  rwa [← eq_empty_of_subset_empty (hB'.subset_ground.trans_eq h)]
+
+@[simp] theorem emptyOn_dual_eq : (emptyOn α)﹡ = emptyOn α := by
+  rw [← ground_eq_empty_iff]; rfl
+
+@[simp] theorem restrict_to_empty (M : Matroid α) : M ↾ (∅ : Set α) = emptyOn α := by
+  simp [← ground_eq_empty_iff]
+
+theorem eq_emptyOn_or_nonempty (M : Matroid α) : M = emptyOn α ∨ Matroid.Nonempty M := by
+  rw [← ground_eq_empty_iff]
+  exact M.E.eq_empty_or_nonempty.elim Or.inl (fun h ↦ Or.inr ⟨h⟩)
+
+theorem ground_nonempty_iff (M : Matroid α) : M.E.Nonempty ↔ M.Nonempty :=
+  ⟨fun h ↦ ⟨h⟩, fun ⟨h⟩ ↦ h⟩
+
+theorem eq_emptyOn [IsEmpty α] (M : Matroid α) : M = emptyOn α := by
+  rw [← ground_eq_empty_iff]
+  exact M.E.eq_empty_of_isEmpty
+
+instance finite_emptyOn (α : Type*) : (emptyOn α).Finite :=
+  ⟨finite_empty⟩
+
+end EmptyOn
+
+section LoopyOn
+
+/-- The `Matroid α` with ground set `E` whose only base is `∅` -/
+def loopyOn (E : Set α) : Matroid α := (emptyOn α ↾ E)
+
+@[simp] theorem loopyOn_ground (E : Set α) : (loopyOn E).E = E := rfl
+
+@[simp] theorem loopyOn_empty (α : Type*) : loopyOn (∅ : Set α) = emptyOn α := by
+  rw [← ground_eq_empty_iff, loopyOn_ground]
+
+@[simp] theorem loopyOn_indep_iff : (loopyOn E).Indep I ↔ I = ∅ := by
+  simp only [loopyOn, restrict_indep_iff, emptyOn_indep_iff, and_iff_left_iff_imp]
+  rintro rfl; apply empty_subset
+
+@[simp] theorem eq_loopyOn_iff : (M = loopyOn E) ↔ M.E = E ∧ ∀ X ⊆ M.E, M.Indep X → X = ∅ := by
```

Comment:
I wouldn't make this one be a `@[simp]` lemma. The RHS isn't simpler, and it also duplicates `M`.